src/jit/codegenxarch.cpp

```diff
@@ -7397,30 +7397,106 @@ void CodeGen::genPutArgStk(GenTreePtr treeNode)
     }
     else if (data->OperGet() == GT_FIELD_LIST)
     {
-        GenTreeFieldList* fieldList = data->AsFieldList();
-        // TODO-X86-CQ: If there are no gaps or padding in the layout, we can use a series of pushes.
-        // For now, always make space for the struct, and store each field at its offset.
+        GenTreePutArgStk* const putArgStk = treeNode->AsPutArgStk();
 
-        unsigned structSize = treeNode->AsPutArgStk()->getArgSize();
-        m_pushStkArg        = false;
-        inst_RV_IV(INS_sub, REG_SPBASE, structSize, EA_PTRSIZE);
-        genStackLevel += structSize;
-        for (; fieldList != nullptr; fieldList = fieldList->Rest())
+        GenTreeFieldList* const fieldList = data->AsFieldList();
+        assert(fieldList != nullptr);
+
+        m_pushStkArg      = false;
+        const int argSize = putArgStk->getArgSize();
+        assert((argSize % TARGET_POINTER_SIZE) == 0);
+
+        // Set the current field offset to the size of the struct arg (which must be a multiple of the target pointer
+        // size).
+        int currentOffset = argSize;
+
+        // If there are no GC references in this field list and the argument is either larger than 16 bytes or the
+        // argument contains padding, we will pre-adjust the stack pointer and use stores instead of pushes.
+        const bool preAdjustStack =
+            (putArgStk->gtNumberReferenceSlots == 0) &&
+            ((argSize > 16) || (putArgStk->gtPutArgStkKind != GenTreePutArgStk::PutArgStkKindRepInstr));
+        if (preAdjustStack)
```

Comment:
Right, that's what I meant by "enhancing the logic as needed". And I think there's actually some value in having just the logic that adjusts the stack between arguments in this method - I think it would even more clearly make that a "special case".

---
In reply to: [82645034](https://github.com/dotnet/coreclr/pull/7537#discussion_r82645034) [](ancestors = 82645034)